File: app/src/main/java/com/nfcalarmclock/NacColorPickerDialog.java
package com.nfcalarmclock;

import android.app.AlertDialog;
import android.content.Context;
import android.content.res.ColorStateList;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.text.Editable;
import android.text.InputType;
import android.text.TextWatcher;
import android.view.KeyEvent;
import android.view.MotionEvent;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;

/**
 * Handle displaying the color picker dialog.
 */
public class NacColorPickerDialog
	extends NacDialog
	implements TextView.OnEditorActionListener,
		TextWatcher,
		View.OnTouchListener,
		NacDialog.OnShowListener
{

	/**
	 * Color picker.
	 */
	private NacColorPicker mColorPicker;

	/**
	 * Color example.
	 */
	private ImageView mColorExample;

	/**
	 * Color hex value.
	 */
	private EditText mEditText;

	/**
	 */
	public NacColorPickerDialog()
	{
		super(R.layout.dlg_color_picker);
		addOnShowListener(this);
	}

	/**
	 * Ensure the text always starts with "#".
	 *
	 * Note: This is required to implement TextWatcher.
	 */
	@Override
	public void afterTextChanged(Editable s)
	{
		if (!s.toString().startsWith("#"))
		{
			s.insert(0, "#");
		}
	}

	/**
	 * Note: This is required to implement TextWatcher.
	 */
	@Override
	public void beforeTextChanged(CharSequence seq, int start, int count, int after)
	{
	}

	/**
	 * @return True if can parse the color, and False otherwise.
	 */
	public boolean canParseColor(String name)
	{
		try
		{
			int color = Color.parseColor(name);
			return true;
		}
		catch (IllegalArgumentException e)
		{
			return false;
		}
	}

	/**
	 * @return The color of the edit text box.
	 */
	public int getColor()
	{
		return this.mColorPicker.getColor();
	}

	/**
	 * @return The EditText color.
	 */
	public String getEditTextColor()
	{
		return this.mEditText.getText().toString();
	}

	/**
	 * @return The last chosen color of the dialog.
	 */
	public String getHexColor()
	{
		return this.mColorPicker.getHexColor();
	}

	/**
	 * Check if valid hex string was input into the EditText.
	 */
	public boolean isHexString(String name)
	{
		for (int i=1; i < name.length(); i++)
		{
			if (Character.digit(name.charAt(i), 16) == -1)
			{
				return false;
			}
		}
		return true;
	}

	/**
	 * Build the dialog.
	 */
	@Override
	public void onBuildDialog(Context context, AlertDialog.Builder builder)
	{
		NacSharedConstants cons = new NacSharedConstants(context);

		builder.setTitle(cons.getTitleColor());
		setPositiveButton(cons.getActionOk());
		setNegativeButton(cons.getActionCancel());
	}

	/**
	 * Close the keyboard when the user hits enter.
	 */
	@Override
	public boolean onEditorAction(TextView tv, int action, KeyEvent event)
	{
		//if ((event == null) && (action == EditorInfo.IME_ACTION_DONE))
		if ((event != null) || (action != EditorInfo.IME_ACTION_DONE))
		{
			return false;
		}

		Context context = tv.getContext();
		NacSharedConstants cons = new NacSharedConstants(context);
		String name = this.getEditTextColor();

		if (!this.isHexString(name) || !this.canParseColor(name))
		{
			NacUtility.quickToast(context, cons.getErrorMessageSelectColor());
			return false;
		}

		int color = Color.parseColor(name);
		this.mColorPicker.setColor(color);
		this.mColorExample.setColorFilter(color, PorterDuff.Mode.SRC);
		closeKeyboard();
		return true;
	}

	/**
	 * Setup the views when the dialog is shown.
	 */
	@Override
	public void onShowDialog(NacDialog dialog, View root)
	{
		Context context = root.getContext();
		NacSharedPreferences shared = new NacSharedPreferences(context);
		this.mColorPicker = root.findViewById(R.id.color_picker);
		this.mColorExample = root.findViewById(R.id.color_example);
		this.mEditText = root.findViewById(R.id.color_value);

		this.mEditText.addTextChangedListener(this);
		this.mEditText.setText(this.getHexColor());
		this.mEditText.setOnEditorActionListener(this);
		this.mEditText.setRawInputType(InputType.TYPE_CLASS_TEXT);
		this.mEditText.setImeOptions(EditorInfo.IME_ACTION_DONE);
		this.mEditText.setBackgroundTintList(ColorStateList.valueOf(shared.getThemeColor()));
		this.mColorPicker.setOnTouchListener(this);
		this.mColorExample.setColorFilter(this.getColor(), PorterDuff.Mode.SRC);
	}

	/**
	 * Note: This is required to implement TextWatcher.
	 */
	@Override
	public void onTextChanged(CharSequence seq, int start, int before, int count)
	{
	}

	/**
	 * Capture touch events on the color picker.
	 */
	@Override
	public boolean onTouch(View view, MotionEvent event)
	{
		view.onTouchEvent(event);

		int color = this.getColor();
		String hex = this.getHexColor();

		this.mColorExample.setColorFilter(color, PorterDuff.Mode.SRC);
		this.mEditText.setText(hex);

		if (event.getAction() == MotionEvent.ACTION_UP)
		{
			view.performClick();
		}

		return true;
	}

	/**
	 * Set the color.
	 */
	public void setColor(int color)
	{
		this.mColorPicker.setColor(color);
		this.mColorExample.setColorFilter(color, PorterDuff.Mode.SRC);

		String hex = this.mColorPicker.getHexColor();

		this.mEditText.setText(hex);
	}

}

File: app/src/main/kotlin/com/nfcalarmclock/view/colorpicker/NacColorPickerDialog.kt
package com.nfcalarmclock.view.colorpicker

import android.app.AlertDialog
import android.app.Dialog
import android.content.Context
import android.content.DialogInterface
import android.graphics.Color
import android.graphics.PorterDuff
import android.os.Build
import android.os.Bundle
import android.view.MotionEvent
import android.view.View
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputMethodManager
import android.widget.ImageView
import com.google.android.material.textfield.TextInputEditText
import com.google.android.material.textfield.TextInputLayout
import com.nfcalarmclock.R
import com.nfcalarmclock.shared.NacSharedPreferences
import com.nfcalarmclock.view.dialog.NacDialogFragment
import com.nfcalarmclock.view.setupInputLayoutColor

/**
 * Handle displaying the color picker dialog.
 */
class NacColorPickerDialog
	: NacDialogFragment(),
	View.OnTouchListener
{

	/**
	 * Listener for when a color is selected.
	 */
	fun interface OnColorSelectedListener
	{
		fun onColorSelected(color: Int)
	}

	/**
	 * Listener for when the default color is selected.
	 */
	fun interface OnDefaultColorSelectedListener
	{
		fun onDefaultColorSelected(dialog: NacColorPickerDialog)
	}

	/**
	 * Initial color to show.
	 */
	var initialColor: Int = 0

	/**
	 * Color picker.
	 */
	private lateinit var colorPicker: NacColorPicker

	/**
	 * Example of the color.
	 */
	private lateinit var exampleColor: ImageView

	/**
	 * Input layout for the hex color.
	 */
	private lateinit var inputLayout: TextInputLayout

	/**
	 * Edit text for the hex color.
	 */
	private lateinit var editText: TextInputEditText

	/**
	 * The color of the edit text box.
	 */
	var color: Int
		get() = colorPicker.color
		set(value)
		{
			// Select the color of the color picker. This will set the HSV,
			// move the color selector, and redraw the shader
			colorPicker.selectColor(value)
		}

	/**
	 * The EditText color.
	 */
	private val editTextColor: String
		get() = editText.text.toString()

	/**
	 * Listener for when the color is selected.
	 */
	var onColorSelectedListener: OnColorSelectedListener? = null

	/**
	 * Listener for when the default color is selected.
	 */
	var onDefaultColorSelectedListener: OnDefaultColorSelectedListener? = null

	/**
	 * Check if can parse the color.
	 *
	 * @return True if can parse the color, and False otherwise.
	 */
	private fun canParseColor(name: String): Boolean
	{
		// Attempt to parse the color
		return try
		{
			Color.parseColor("#$name")
			true
		}
		// Unable to parse the color
		catch (e: IllegalArgumentException)
		{
			false
		}
	}

	/**
	 * Close the keyboard.
	 */
	@Suppress("deprecation")
	private fun closeKeyboard()
	{
		val context = requireContext()

		// Check the Android version. It needs to be done differently depending on
		// the version due to deprecation
		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S)
		{
			val inputManager = context.getSystemService(InputMethodManager::class.java)
			val token = dialog!!.currentFocus?.windowToken

			// Hide the keyboard
			inputManager.hideSoftInputFromWindow(token, 0)
		}
		// API < 31
		else
		{
			val inputManager = context.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
			//val flags = InputMethodManager.HIDE_IMPLICIT_ONLY

			// Hide the keyboard
			//inputManager.toggleSoftInput(flags, 0)
			inputManager.toggleSoftInput(0, 0)
		}
	}

	/**
	 * Check if valid hex string was input into the EditText.
	 *
	 * @return True if it is a valid hex string, and False otherwise.
	 */
	private fun isHexString(name: String): Boolean
	{
		// Iterate over each character in the hex string
		name.forEach { char ->

			// Convert the character to base 16
			val digit = char.digitToIntOrNull(16) ?: -1

			// Unable to convert the character so this is not a hex string
			if (digit == -1)
			{
				return false
			}

		}

		// Success. This is a hex string
		return true
	}

	/**
	 * Called when the dialog is created.
	 */
	override fun onCreateDialog(savedInstanceState: Bundle?): Dialog
	{
		// Setup the shared preferences
		setupSharedPreferences()

		// Create the dialog
		return AlertDialog.Builder(requireContext())
			.setTitle(R.string.title_select_color)
			.setPositiveButton(R.string.action_ok) { _, _ ->

				// Call the listener
				onColorSelectedListener?.onColorSelected(color)

			}
			.setNegativeButton(R.string.action_cancel, null)
			.setNeutralButton(R.string.action_default, null)
			.setView(R.layout.dlg_color_picker)
			.create()
	}

	/**
	 * Called when the fragment is resumed.
	 */
	override fun onResume()
	{
		// Super
		super.onResume()

		// Setup the member variables
		colorPicker = dialog!!.findViewById(R.id.color_picker)
		exampleColor = dialog!!.findViewById(R.id.color_example)
		editText = dialog!!.findViewById(R.id.color_edittext)
		inputLayout = dialog!!.findViewById(R.id.color_input_layout)

		// Set the initial color
		color = initialColor
		updateExampleAndHexColors(initialColor)

		// Setup the views
		colorPicker.setOnTouchListener(this)
		setupEditText()
		setupNeutralButtonOnClickListener()
	}

	/**
	 * Capture touch events on the color picker.
	 */
	override fun onTouch(view: View, event: MotionEvent): Boolean
	{
		// Call the touch event
		view.onTouchEvent(event)

		// Update the example color and hex color
		updateExampleAndHexColors(color)

		// Perform click
		if (event.action == MotionEvent.ACTION_UP)
		{
			view.performClick()
		}

		return true
	}

	/**
	 * Setup the edit text color.
	 */
	private fun setupEditText()
	{
		// Get the context
		val context = requireContext()

		// Setup the color
		inputLayout.setupInputLayoutColor(context, sharedPreferences!!)

		// Set the keyboard action listener so that the keyboard closes when the user
		// hits enter
		editText.setOnEditorActionListener { _, action, event ->

			// Check if a keyboard action occurred
			if ((event != null) || (action != EditorInfo.IME_ACTION_GO))
			{
				return@setOnEditorActionListener false
			}

			// Check if the color is not hex or it cannot be parsed
			if (!isHexString(editTextColor) || !canParseColor(editTextColor))
			{
				// Show error hint message
				inputLayout.setError(getString(R.string.error_message_select_color))
				return@setOnEditorActionListener false
			}

			// Set the new color on the color picker, example color, and hex color
			color = Color.parseColor("#$editTextColor")
			updateExampleAndHexColors(color)

			// Close the keyboard
			closeKeyboard()
			return@setOnEditorActionListener true
		}
	}

	/**
	 * Setup neutral button on click listener.
	 */
	private fun setupNeutralButtonOnClickListener()
	{
		// Get the neutral button
		val alertDialog = dialog as AlertDialog
		val neutralButton = alertDialog.getButton(DialogInterface.BUTTON_NEUTRAL)

		// Set the listener
		neutralButton.setOnClickListener {

			// Call the listener
			onDefaultColorSelectedListener?.onDefaultColorSelected(this)

			// Update the example color and hex color
			updateExampleAndHexColors(color)

		}
	}

	/**
	 * Update the example and hex colors.
	 */
	private fun updateExampleAndHexColors(newColor: Int)
	{
		// Set the example color
		exampleColor.setColorFilter(newColor, PorterDuff.Mode.SRC)

		// Set the hex color in the edit text
		val hexColor = String.format("%06X", (0xFFFFFF and newColor))

		editText.setText(hexColor)

		// Clear the error hint message, if it was set
		inputLayout.error = null
	}

	companion object
	{

		/**
		 * Tag for the class.
		 */
		const val TAG = "NacColorPickerDialog"

	}

}